Fixes #1649. Exclusion rules and other cached resolution fixes

eed3si9n opened the pull request:
Fixes #1649.

eed3si9n committed:
Fixes #1649. Adds exclusion rules into SHA-1 of minigraph path

lightbend-cla-validator commented:
Hi @eed3si9n,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.typesafe.com/contribute/cla](http://www.typesafe.com/contribute/cla)

jsuereth commented on ivy/src/main/scala/sbt/ivyint/CachedResolutionResolveEngine.scala:
```diff
@@ -59,11 +59,28 @@ private[sbt] class CachedResolutionResolveCache() {
+      def excludeRuleString(rule: ExcludeRule): String =
+        s"""Exclude(${rule.getId},${rule.getConfigurations.mkString(",")},${rule.getMatcher})"""
+      def includeRuleString(rule: IncludeRule): String =
+        s"""Include(${rule.getId},${rule.getConfigurations.mkString(",")},${rule.getMatcher})"""
```
Hmm..... this feels less consistent....   Are the toStrings of rule.getMatcher pretty consistent in general?

eed3si9n replied:
Yes. Ivy, like many Java libraries, have nice `toString()` normally. :neutral_face:

jsuereth commented:
Generally LGTM

eed3si9n committed:
Fixes #1639. Fixes cached resolution interacting with force()

When conflicts are found for a given module, a forced one
is selected before conflict manager kicks in.
The problem is that DependencyDescriptor seems to mark transitive
forced dependency as forced as well,
so the actual forced dependency are sometimes not prioritized.
To work around this, I’ve introduced a mixin called
SbtDefaultDependencyDescriptor, which carries around ModuleID to detect
direct dependencies.

eed3si9n commented:
Since this was not merged yet, I'm stacking another bug fix on top.

jsuereth commented:
LGTM, again

eed3si9n committed:
Fixes #1660

Provide overloads for transitiveScratch and updateClassifiers that work
with IvyActions.updateEither.

eed3si9n committed:
Make sure IvyAction.update works

eed3si9n committed:
Use correct stats for cached resolution

eed3si9n committed:
Make sure all files exist in minigraph.

eed3si9n commented:
More fixes were added yesterday. Tests failing, but not sure if it's me or Travis.

eed3si9n committed:
Implement configuration.missingOk to fix updateClassifiers

updateClassifiers sets missingOk to true.
cached resolution wasn’t respecting the looseness.

eed3si9n commented:
@jsuereth so you were right about Travis error being legit. Good job, Travis/scripted!
`updateClassifiers` sets a flag called `missingOk`, which allows Ivy to explorer around to grab JARs, but not let the unresolved error propagate as error. I needed to port the behavior for cached resolution.

jsuereth commented:
LGTM
